### PR TITLE
feat: add support for fantom

### LINKED ISSUE #4873: [Demand] Introduce Fantom Chain
## Catalogue

- [ ] UI/UX
- [ ] Mask Core Feature
- [x] Mask Plugin Feature
- [ ] Security/Cryptography Feature
- [ ] Other
  - /_List it here_/

## Part 1: Introduction and Background

| Property | Description |
| --- | ---- |
| DEX |  https://spookyswap.finance/swap, https://github.com/SpookySwap (An uniswap-v2 like DEX.) | 
| ChainId |  Fantom Opera -  `250` / Fantom Testnet - `4002` | 
| RPC |  https://rpc.ftm.tools/ | 
| CoinGecko ID | `"fantom"` | 
| CMC ID | `"3513"` | 
| Debank ID | `"fantom"` | 
| OpenSea |  Not Supported | 
| Explorer |  https://ftmscan.com/ | 

## Notes

+ [Fantom](https://fantom.foundation/)
+ [Documentation](https://docs.fantom.foundation/)
+ [FTMScan](https://ftmscan.com/)

### PR DESCRIPTION
**Pull request by UncleBill**
resolves #4873 

## Need to confirm:

+ [x] `Debank`: Check the asset list on the dashboard page.
+ [x] `coingecko` & `CMC`: Check trending view when hover $FTM token. 
+ [x] Trade with SushiSwap on Fantom.
+ [x] Send a transaction on Fantom and check the explorer is working.
+ [x] Ensure the gas estimate dialog is working.

### CODE REVIEW

File: packages/mask/src/plugins/EVM/constants.ts [hunk 1]
```diff
@@ -127,6 +127,16 @@ export const PLUGIN_NETWORKS: Web3Plugin.NetworkDescriptor[] = [
         iconColor: 'rgb(73, 169, 166)',
         isMainnet: true,
     },
+    {
+        ID: `${PLUGIN_ID}_fantom`,
+        networkSupporterPluginID: PLUGIN_ID,
+        chainId: ChainId.Fantom,
+        type: NetworkType.Fantom,
+        name: 'Fantom',
+        icon: new URL('./assets/fantom.svg', import.meta.url),
```
Comment (guanbinrui) [inline packages/mask/src/plugins/EVM/constants.ts:136]: Use a PNG image.

File: packages/mask/src/plugins/Trader/constants/0x.ts [hunk 2]
```diff
@@ -6,6 +6,7 @@ export const ZRX_BASE_URL: Record<NetworkType, string> = {
     [NetworkType.Polygon]: 'https://polygon.api.0x.org/',
     [NetworkType.Arbitrum]: 'https://arbitrum.api.0x.org/',
     [NetworkType.xDai]: 'https://xdai.api.0x.org/',
+    [NetworkType.Fantom]: 'https://fantom.api.0x.org/',
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/constants/0x.ts:9]: Tested?

File: packages/mask/src/plugins/Wallet/SNSAdaptor/ConnectWalletDialog/index.tsx [hunk 3]
```diff
@@ -122,7 +122,8 @@ export function ConnectWalletDialog(props: ConnectWalletDialogProps) {
                 // recheck
                 const chainIdHex = await Services.Ethereum.getChainId(overrides)
                 if (Number.parseInt(chainIdHex, 16) !== expectedChainId) throw new Error('Failed to switch chain.')
-            } catch {
+            } catch (err) {
+                console.log({ err })
                 throw new Error(`Make sure your wallet is on the ${resolveNetworkName(networkType)} network.`)
             }
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Wallet/SNSAdaptor/ConnectWalletDialog/index.tsx:128]:
```suggestion
            } catch {
                throw new Error(`Make sure your wallet is on the ${resolveNetworkName(networkType)} network.`)
            }
```

File: packages/mask/src/plugins/Wallet/SNSAdaptor/ConnectWalletDialog/index.tsx [hunk 4]
```diff
@@ -122,7 +122,7 @@ export function ConnectWalletDialog(props: ConnectWalletDialogProps) {
                 // recheck
                 const chainIdHex = await Services.Ethereum.getChainId(overrides)
                 if (Number.parseInt(chainIdHex, 16) !== expectedChainId) throw new Error('Failed to switch chain.')
-            } catch {
+            } catch (err) {
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Wallet/SNSAdaptor/ConnectWalletDialog/index.tsx:125]:
```suggestion
            } catch {
```

File: packages/web3-shared/evm/types/index.ts [hunk 5]
```diff
@@ -77,6 +80,7 @@ export enum NetworkType {
     Polygon = 'Polygon',
     Arbitrum = 'Arbitrum',
     xDai = 'xDai',
+    Fantom = 'Fantom',
```
Comment (guanbinrui) [inline packages/web3-shared/evm/types/index.ts:83]: sync it to `packages/public-api/src/web.ts`

**Comment by guanbinrui** [comment]
@UncleBill New updates on develop branch.

**Comment by guanbinrui** [comment]
I've written a checklist for CELO. Please also confirm those items in that list for the Fantom chain.

https://github.com/DimensionDev/Maskbook/pull/5052#issuecomment-986258745

File: packages/mask/src/plugins/Wallet/SNSAdaptor/ConnectWalletDialog/index.tsx [hunk 6]
```diff
@@ -125,7 +125,7 @@ export function ConnectWalletDialog(props: ConnectWalletDialogProps) {
...
                 throw new Error(`Make sure your wallet is on the ${resolveNetworkName(networkType)} network.`)
```
Comment (Jack-Works) [inline packages/mask/src/plugins/Wallet/SNSAdaptor/ConnectWalletDialog/index.tsx:129]:
```suggestion
                throw new Error(`Make sure your wallet is on the ${resolveNetworkName(networkType)} network.`, { cause: err })
```

File: packages/mask/src/manifest.json [hunk 7]
```diff
@@ -10,6 +10,6 @@
     "browser_action": { "default_popup": "popups.html" },
     "homepage_url": "https://mask.io",
     "description": "The portal to the new & open Internet. Send encrypted message and decentralized Apps right on top of social networks.",
-    "web_accessible_resources": ["js/*", "*.png"],
+    "web_accessible_resources": ["js/*", "*.png", "*.svg"],
```
Comment (Jack-Works) [inline packages/mask/src/manifest.json:13]: revert this since you no longer need it

**Comment by UncleBill** [comment]
Check list based on https://github.com/DimensionDev/Maskbook/pull/5052#issuecomment-986258745

+ [ ] Check if the asset list on the dashboard page is working when you choose Fantom as the network.
    + Setup Fantom for DEBANK API.
+ [ ] Check if the trending view is working. Try to hover a Fantom token in any tweets.
    + Setup Fantom for Coingeck API.
    + Setup Fantom for CMC API.

image

+ [ ] Check if the gas estimate dialog is working.
    + Setup Fantom for Coingeck API.

 Goto `chrome-extension://jkoeaghipilijlahjplgbfiocjhldnap/dashboard.html#/wallets/transfer` and check the estimated USD value is working.

image

+ [ ] Trade with the DEX on Fantom.
     + Integrate a DEX for Fantom.

+ [ ] Send a transaction and check if the explorer links are working.

File: packages/mask/src/plugins/Trader/trader/uniswap/useAllCommonPairs.ts [hunk 8]
```diff
@@ -29,7 +29,7 @@ export function useAllCurrencyCombinations(tradeProvider: TradeProvider, currenc
         [bases],
     )
 
-    return useMemo(() => {
+    const combinations = useMemo(() => {
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/trader/uniswap/useAllCommonPairs.ts:32]: Revert it.

File: packages/mask/src/plugins/Trader/trader/uniswap/usePairs.ts [hunk 9]
```diff
@@ -19,15 +19,17 @@ export enum PairState {
 
 export type TokenPair = [Token, Token]
 
+const EMPTY_LIST: never[] = []
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/trader/uniswap/usePairs.ts:22]: Not necessary.

File: packages/web3-shared/evm/hooks/useAssets.ts [hunk 10]
```diff
@@ -1,12 +1,13 @@
...
 import { useWallet } from './useWallet'
 import { useNativeTokenDetailed } from './useNativeTokenDetailed'
 import { useAssetsFromChain } from './useAssetsFromChain'
 import { useAssetsFromProvider } from './useAssetsFromProvider'
-import { useCallback } from 'react'
+import { useCallback, useMemo } from 'react'
 import { useAssetsMerged } from './useAssetsMerged'
 import { formatEthereumAddress } from '../utils'
 
+const emptyList: Asset[] = []
```
Comment (guanbinrui) [inline packages/web3-shared/evm/hooks/useAssets.ts:10]: Not necessary.

File: packages/web3-shared/evm/hooks/useAssetsFromChain.ts [hunk 11]
```diff
@@ -6,7 +6,9 @@ import { useBalance } from './useBalance'
 import { getChainDetailed } from '../utils'
 import { useProviderType } from './useProviderType'
 import { useBalances } from './useBalances'
+import { useMemo } from 'react'
 
+const emptyList = [] as const
```
Comment (guanbinrui) [inline packages/web3-shared/evm/hooks/useAssetsFromChain.ts:11]: Not necessary.

File: packages/web3-shared/evm/hooks/useAssetsMerged.ts [hunk 12]
```diff
@@ -3,7 +3,9 @@ import { useChainId } from './useChainId'
 import { uniqBy } from 'lodash-unified'
 import { formatEthereumAddress, makeSortAssertFn } from '../utils'
 import { useTokenConstants } from '../constants'
+import { useMemo } from 'react'
 
+const emptyList: Asset[] = []
```
Comment (guanbinrui) [inline packages/web3-shared/evm/hooks/useAssetsMerged.ts:8]: Not necessary.

File: packages/web3-shared/evm/hooks/useMulticall.ts [hunk 13]
```diff
@@ -132,10 +132,14 @@ export function useMulticallCallback(targetChainId?: ChainId, targetBlockNumber?
                     await Promise.all(
                         chunkArray(unresolvedCalls).map(async (chunk) => {
                             // we don't mind the actual block number of the current call
-                            const { returnData } = await multicallContract.methods.multicall(chunk).call(overrides)
-                            returnData.forEach((result, index) =>
-                                setCallResult(chunk[index], result, chainId, blockNumber),
-                            )
+                            try {
```
Comment (guanbinrui) [inline packages/web3-shared/evm/hooks/useMulticall.ts:135]: Revert it.

File: packages/web3-shared/evm/hooks/useAssetsMerged.ts (hunk 12 above)
Comment (guanbinrui) [inline packages/web3-shared/evm/hooks/useAssetsMerged.ts:8]:
```suggestion
const EMPTY_LIST: Asset[] = []
```

File: packages/web3-shared/evm/hooks/useAssets.ts (hunk 10 above)
Comment (UncleBill) [reply packages/web3-shared/evm/hooks/useAssets.ts:10]: Yes, I plan to replace all inline array in the code base with a stable array, that will reduce re-renderings.

Comment (guanbinrui) [reply packages/web3-shared/evm/hooks/useAssets.ts:10]: I see. Better to name them `EMPTY_LIST`.

Comment (UncleBill) [reply packages/web3-shared/evm/hooks/useAssets.ts:10]: I will.

Comment (UncleBill) [reply packages/web3-shared/evm/hooks/useAssets.ts]: Will refactor EMPTY_LIST in another PR.

**Comment by UncleBill** [comment]
Please help to verify @DimensionDev/qa

**Review by nuanyang233 — DISMISSED** [review]

File: packages/mask/src/components/shared/ApplicationBoard.tsx [hunk 14]
```diff
@@ -301,7 +309,15 @@ export function ApplicationBoard({ secondEntries, secondEntryChainTabs }: MaskAp
                             () => {},
                         ),
                     ],
-                    [ChainId.Mainnet, ChainId.BSC, ChainId.Matic, ChainId.Arbitrum, ChainId.xDai, ChainId.Celo],
+                    [
```
Comment (septs) [inline packages/mask/src/components/shared/ApplicationBoard.tsx:312]: may be can extract as multi-presets.

File: packages/dashboard/src/pages/Labs/components/SwapSettingDialog.tsx [hunk 15]
```diff
@@ -89,6 +91,12 @@ export default function SwapSettingDialog({ open, onClose }: SettingDialogProps)
             options: celoOptions,
             onChange: (value: string) => Services.Settings.setCeloNetworkTradeProvider(Number.parseInt(value, 10)),
         },
+        {
+            legend: t.labs_settings_swap_network({ network: 'Fantom' }),
+            value: useFantomNetworkTradeProvider(),
+            options: fantomOptions,
+            onChange: (value: any) => Services.Settings.setFantomNetworkTradeProvider(+value),
```
Comment (guanbinrui) [inline packages/dashboard/src/pages/Labs/components/SwapSettingDialog.tsx:98]: Avoid `any`.

Comment (UncleBill) [reply packages/dashboard/src/pages/Labs/components/SwapSettingDialog.tsx]: Fixed. Thanks.

File: packages/web3-constants/evm/token.json [hunk 16]
```diff
@@ -1,4 +1,19 @@
 {
+    "ZERO_ADDRESS": {
```
Comment (guanbinrui) [inline packages/web3-constants/evm/token.json:2]: It has been removed.

**Review by guanbinrui — APPROVED** [review]

File: packages/web3-constants/evm/token.json (hunk 16 above)
Comment (UncleBill) [reply packages/web3-constants/evm/token.json]: Removed.

**Comment by UncleBill** [comment]
@DimensionDev/qa please help to verify, thanks!

File: packages/mask/src/components/shared/ApplicationBoard.tsx (hunk 14 above)
Comment (UncleBill) [reply packages/mask/src/components/shared/ApplicationBoard.tsx:312]: They are used only in this file, not very common, and extracting will decrease the flexibility. We can do it anytime when we need to.

**Review by Jack-Works — APPROVED** [review]

**Review by guanbinrui — APPROVED** [review]

**Pull request merged**